### PR TITLE
feat(components/drawer): adds drawer component to the library

### LINKED ISSUE #340: [components] Drawer
We need to implement the Drawer Label component as described in the core library.
https://flowbite.com/docs/components/drawer/

**Describe the solution you'd like**

PR must have:

- Component (w/ Theme support)
- Storybook
- Documentation
- Unit test

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by areski**
## Description

This is my first Typescript PR, so be gentle here :D

WIP implementation of `Drawer` https://flowbite-react.com/modal

There is a setting for `backdrop` and `placement`, but `scrolling` and `swipeable edge` is not implemented.

**To be noted:**
1. The drawer transition is not working, I will need some direction on how to correct this.
2. `backdrop` is not working when clicking outside of the component, any pointer would be welcome

Looking for feedback, thanks!

Fixes #340

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change contains documentation update

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

File: src/lib/theme/default.ts [hunk 1]
```diff
@@ -562,6 +562,48 @@ const theme: FlowbiteTheme = {
       popup: 'border-t',
     },
   },
+  drawer: {
+    root: {
+      // base: 'fixed top-0 right-0 left-0 z-50 h-modal overflow-y-auto overflow-x-hidden md:inset-0 md:h-full',
```
Comment (rluders) [inline src/lib/theme/default.ts:567]: Remove this line. :)

File: src/lib/components/Flowbite/FlowbiteTheme.ts [hunk 2]
```diff
@@ -136,6 +138,13 @@ export interface FlowbitePositions {
   'center-right': string;
 }
 
+export interface FlowbitePlacements {
+  left: { on: string; off: string };
```
Comment (rluders) [inline src/lib/components/Flowbite/FlowbiteTheme.ts:142]: Isn't better to create an interface to `on: string; off: string;`?

File: src/lib/components/Drawer/Drawer.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+  useKeyDown('Escape', () => {
+    if (backdrop && onClose) {
+      onClose();
+    }
+  });
+
+  const handleOnClick = (e: MouseEvent<HTMLDivElement>) => {
+    console.log('...handleOnClick');
+    console.log('backdrop:', backdrop);
```
Comment (rluders) [inline src/lib/components/Drawer/Drawer.tsx:97]: Remove the `console.log` lines here.

**Review by rluders — CHANGES_REQUESTED** [review]
@areski great job. I didn't have the time to test it properly yet. I leave some small comments requesting some changes.

File: src/lib/components/Flowbite/FlowbiteTheme.ts (hunk 2 above)
Comment (areski) [reply src/lib/components/Flowbite/FlowbiteTheme.ts]: Ok

**Comment by areski** [comment]
Thanks, I pushed a fix based on your comments.

**Comment by rluders** [comment]
@areski I tested it today and there are a few other things that I noticed:

1. Backdrop isn't working on the docs page. (ref: https://flowbite.com/docs/components/drawer/#backdrop)
2. I can scroll the page, by default, it shouldn't be possible (ref: https://flowbite.com/docs/components/drawer/#body-scrolling)
3. Animation is missing. (ref: https://flowbite.com/docs/components/drawer/#default-drawer)

**Comment by areski** [comment]
@rluders I have mentioned some of your points in the ticket description.

Unfortunately my priority have shifted.
Can someone take over this PR?

**Comment by rluders** [comment]
Oh, I didn't notice that on the ticket. Anyway, sure, @areski... I can co-author this one, I guess. Thank you!

**Comment by areski** [comment]
Thanks, much appreciated

**Comment by xiaohuilam** [comment]
Hello @areski, I got this BUG when I try your patch: if set `backdrop={true}` when i click the empty area outside of drawer, it triggers nothing, but if i click empty area inside it, it closes. I think this is wrong proceed of backdrop.

.tsx code to reproduce:
```tsx
  const [isDrawerOpen, isDrawerOpenChange] = React.useState(false);

  return (
    <Button onClick={() => isDrawerOpenChange(true)}>
      Show Drawer
    </Button>
    <Drawer show={isDrawerOpen} backdrop={true} placement="right" className="w-1/2" onClose={() => isDrawerOpenChange(false)}>
      <Drawer.Header>
        修改
      </Drawer.Header>
      <Drawer.Body>
        aaa
      </Drawer.Body>
    </Drawer>
  );
```

**Comment by xiaohuilam** [comment]
@areski 


Because of [upstream changes](https://github.com/themesberg/flowbite-react/commit/6e2b0911ce566b3ec91ebd775fbfde7047669f36)

You'd update these `src/lib/components/Drawer/Drawer.tsx`'s code:

https://github.com/themesberg/flowbite-react/blob/51f7ae81a51b693c32aa7c59bae8d484d5c9555d/src/lib/components/Drawer/Drawer.tsx#L89-L93


into

```
  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === 'Escape' && onClose) {
        onClose();
      }
    };

    document.addEventListener('keydown', handleKeyDown);

    return () => {
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, [onClose]);

  const handleOnClick = (e: MouseEvent<HTMLDivElement>) => {
    if (backdrop && e.target === e.currentTarget && onClose) {
      onClose();
    }
  };
```

And rename `src/lib/components/Drawer` into `src/components/Drawer`.

**Comment by rluders** [comment]
@xiaohuilam don't you want to take the commits on this PR and create another PR with the changes? :)

**Comment by tulup-conner** [comment]
@xiaohuilam Let us know if you're able to take over here, because the OP @areski has already stated recently he's unable to focus on this now. You'd just need to follow [the steps here](https://gist.github.com/wtbarnes/56b942641d314522094d312bbaf33a81) except, instead of force-pushing to their branch, which you won't be able to do, you'd just need to create your own fork and a PR from that.

**Comment by xiaohuilam** [comment]
> @xiaohuilam Let us know if you're able to take over here, because the OP @areski has already stated recently they're unable to focus on this now. You'd just need to follow [the steps here](https://gist.github.com/wtbarnes/56b942641d314522094d312bbaf33a81) except, instead of force-pushing to their branch, which you won't be able to do, you'd just need to create your own fork and a PR from that. If that's complicated, we can take over!

@tulup-conner @rluders I’d like to contribute but I have no experience to modify pull requests created by others. Should I open a new PR from my fork?

**Comment by rluders** [comment]
> > @xiaohuilam Let us know if you're able to take over here, because the OP @areski has already stated recently they're unable to focus on this now. You'd just need to follow [the steps here](https://gist.github.com/wtbarnes/56b942641d314522094d312bbaf33a81) except, instead of force-pushing to their branch, which you won't be able to do, you'd just need to create your own fork and a PR from that. If that's complicated, we can take over!
> 
> @tulup-conner @rluders I’d like to contribute but I have no experience to modify pull requests created by others. Should I open a new PR from my fork?

Sure, you can get this PR for your local fork, and then create another PR based on this one.

**Pull request closed without merging**